### DISCUSSION TITLE
[all] socket 연결과 nginx

### DISCUSSION THREAD

**Discussion by JuneParkCode** [3DPong/transcendence#130] [General]
## 발단
오늘 무슨 이유인지 소켓 연결이 안됐습니다.  아마 nginx 문제라고 생각했었는데 맞습니다...

## 수정
https://github.com/3DPong/transcendence/blob/3f83032fafcc5da20a74686a4d97ce7c976bb6b5/nginx/nginx.conf#L1-L58
위 코드는 수정된 코드입니다.

proxy_pass 과정에서 제대로 연결을 못해주는게 문제입니다.

### api-server 와의 ws 연결
프론트의 경우 ws 연결을 위해서 `handshake path` (프론트에서는 path 지정하는 것) 을 `/api/socket.io` 로 요청을 합니다.
그러나, 현재 api-server 입장에서는 `/api/socket.io` 경로로는 handshake가 불가능합니다. path가 기본값으로 `/socket.io`로 잡혀있으니까요.
이를 해결하는 방법은 두 가지가 있습니다.
1. nginx에서 수정
```nginx
 # socket io 통신용 
         location /api/socket.io { 
             proxy_pass         http://api-server/socket.io; 
             # websocket 통신 
             proxy_http_version 1.1; 
             proxy_set_header Upgrade $http_upgrade; 
             proxy_set_header Connection "upgrade"; 
         } 
```
2. SocketIOModule config 잡기
```typescript
import { Module } from '@nestjs/common';
import { SocketIoModule } from '@nestjs/websockets';
import { AppGateway } from './app.gateway';

@Module({
  imports: [
    SocketIoModule.forRoot({
      path: '/my-custom-path', // set the desired path here
      // other Socket.IO options go here
    }),
  ],
  providers: [AppGateway],
})
export class AppModule {}
```

2번으로 해도 괜찮기는 한데, socket.io 자체에서  proxy_pass를 하는 경우, 아래와 같은 설정이 필요합니다.
```nginx
 proxy_http_version 1.1; 
 proxy_set_header Upgrade $http_upgrade; 
 proxy_set_header Connection "upgrade"; 
```
`/api` 쪽에 다 잡하두는 것도 이상한 것 같아서, 지금 형태로 해도 무방하지 않을까 싶습니다.

### webpack-dev-server 와 ws 연결
이건 조금 다른 문제인데,  그냥 아무것도 설정하지 않으면 다음과 같은 에러를 만납니다. (아마 배포환경에서는 발생하지 않을 문제...)
<img width="434" alt="image" src="https://user-images.githubusercontent.com/81505228/231543538-8e280e89-5809-4f2c-a362-82eda19ae0ab.png">
`webpack-dev-server`는 디버깅&라이브 업데이트를 위해서 소켓 통신을 합니다. 이 통신을 ws를 통해서 하게 되는데, ws handshake path가 바로 `/ws` 입니다.
그래서  앞서 `api-server`에서 해결했던 방식으로 해결할 수 있습니다. `/ws`로 들어올 경우, `front-server` 쪽으로 proxy_pass를 하고, 마찬가지로 프로토콜 업그레이드에 필요한 세팅을 해주면 됩니다.
```nginx
# webpack-dev-server socket 통신용 
location /ws { 
   proxy_pass         http://front-server/ws; 
   proxy_http_version 1.1; 
   proxy_set_header Upgrade $http_upgrade; 
   proxy_set_header Connection "upgrade"; 
} 
```
그게 이 부분입니다.

해당 내용은 3f83032fafcc5da20a74686a4d97ce7c976bb6b5 커밋을 통해서 업데이트 되었습니다.

그리고,
앞으로 테스트는 반드시 `integration-main` 의 파생 브랜치에서 해주셔야, 서로 테스트 환경 동기화하면서 작업하기 편합니다. 
#129  꼭 읽고 해당 방법으로 테스트해주세요.

**Comment by kimminkyeu** [comment]
오늘부터 해당 방법으로 테스트 진행하겠습니다. 빠른 업데이트 감사합니다!
